Add buffer transfer optimization to gpu_to_comp pass

YangleiZouIntel opened the pull request:
1. adds buffer transfer optimization at comp level  https://github.com/plaidml/plaidml/pull/1356#issuecomment-700854644
    OpenCL resnet50 overall inference latency improved from 74ms to 61ms (about 18% improvement) on my local machine with Gen9 . Resnet50 total number of memory transfers reduced from 473 to 230

2. optimized comp dialect representation to help support memory optimization and share comp dialect representation between ocl, vulkan and L0

Example: PLAIDML_DEVICE=vulkan.0 PLAIDML_TARGET=intel_gen bazelisk run plaidml/bridge/keras:backend_test -- TestBackendOps.testConv1d 

Below shows one of the conv1d test with input data shape  <2x3x1xf32>,<1x8x3xf32>, <1x8x1xf32>

Before 
```
func @main(%arg0: !comp.device, %arg1: memref<2x3x1xf32>, %arg2: memref<1x8x3xf32>, %arg3: memref<1x8x1xf32>) {
    %cst = constant 0.000000e+00 : f32
    %c2 = constant 2 : index
    %c4 = constant 4 : index
    %c5 = constant 5 : index
    %c6 = constant 6 : index
    %c7 = constant 7 : index
    %c0 = constant 0 : index
    %c10 = constant 10 : index
    %c3 = constant 3 : index
    %c8 = constant 8 : index
    %c1 = constant 1 : index
    %0 = alloc() : memref<1x10x3xf32>
    %1 = comp.create_execenv %arg0 : (!comp.device) -> !comp.execenv<ocl:0,(11)>
    %2 = comp.alloc %1 : (!comp.execenv<ocl:0,(11)>) -> memref<1x10x3xf32, 11>
    %3 = comp.schedule_write %0 to %2 on %1 : (memref<1x10x3xf32>, memref<1x10x3xf32, 11>, !comp.execenv<ocl:0,(11)>) -> !comp.event<ocl>
    comp.wait %3 : !comp.event<ocl>
    %4 = "comp.schedule_func"(%1) ( {
      "gpu.launch_func"(%c1, %c1, %c1, %c1, %c10, %c3, %2) {kernel = @main_kernel::@main_kernel} : (index, index, index, index, index, index, memref<1x10x3xf32, 11>) -> ()
      "comp.schedule_end"() : () -> ()
    }) : (!comp.execenv<ocl:0,(11)>) -> !comp.event<ocl>
    %5 = comp.schedule_read %0 from %2 on %1 wait for %4 : (memref<1x10x3xf32>, memref<1x10x3xf32, 11>, !comp.execenv<ocl:0,(11)>, !comp.event<ocl>) -> !comp.event<ocl>
    comp.wait %5 : !comp.event<ocl>
    comp.dealloc %1 %2 : (!comp.execenv<ocl:0,(11)>, memref<1x10x3xf32, 11>) -> ()
    comp.destroy_execenv %1 : !comp.execenv<ocl:0,(11)>
    %6 = comp.create_execenv %arg0 : (!comp.device) -> !comp.execenv<ocl:0,(11)>
    %7 = comp.alloc %6 : (!comp.execenv<ocl:0,(11)>) -> memref<1x8x3xf32, 11>
    %8 = comp.schedule_write %arg2 to %7 on %6 : (memref<1x8x3xf32>, memref<1x8x3xf32, 11>, !comp.execenv<ocl:0,(11)>) -> !comp.event<ocl>
    comp.wait %8 : !comp.event<ocl>
    %9 = comp.alloc %6 : (!comp.execenv<ocl:0,(11)>) -> memref<1x10x3xf32, 11>
    %10 = comp.schedule_write %0 to %9 on %6 : (memref<1x10x3xf32>, memref<1x10x3xf32, 11>, !comp.execenv<ocl:0,(11)>) -> !comp.event<ocl>
    comp.wait %10 : !comp.event<ocl>
    %11 = "comp.schedule_func"(%6) ( {
      "gpu.launch_func"(%c1, %c1, %c1, %c1, %c8, %c3, %7, %9) {kernel = @main_kernel_0::@main_kernel} : (index, index, index, index, index, index, memref<1x8x3xf32, 11>, memref<1x10x3xf32, 11>) -> ()
      "comp.schedule_end"() : () -> ()
    }) : (!comp.execenv<ocl:0,(11)>) -> !comp.event<ocl>
    %12 = comp.schedule_read %arg2 from %7 on %6 wait for %11 : (memref<1x8x3xf32>, memref<1x8x3xf32, 11>, !comp.execenv<ocl:0,(11)>, !comp.event<ocl>) -> !comp.event<ocl>
    %13 = comp.schedule_read %0 from %9 on %6 wait for %11 : (memref<1x10x3xf32>, memref<1x10x3xf32, 11>, !comp.execenv<ocl:0,(11)>, !comp.event<ocl>) -> !comp.event<ocl>
    comp.wait %12, %13 : !comp.event<ocl>, !comp.event<ocl>
    comp.dealloc %6 %7 : (!comp.execenv<ocl:0,(11)>, memref<1x8x3xf32, 11>) -> ()
    comp.dealloc %6 %9 : (!comp.execenv<ocl:0,(11)>, memref<1x10x3xf32, 11>) -> ()
    comp.destroy_execenv %6 : !comp.execenv<ocl:0,(11)>
    %14 = comp.create_execenv %arg0 : (!comp.device) -> !comp.execenv<ocl:0,(11)>
    %15 = comp.alloc %14 : (!comp.execenv<ocl:0,(11)>) -> memref<1x8x1xf32, 11>
    %16 = comp.schedule_write %arg3 to %15 on %14 : (memref<1x8x1xf32>, memref<1x8x1xf32, 11>, !comp.execenv<ocl:0,(11)>) -> !comp.event<ocl>
    comp.wait %16 : !comp.event<ocl>
    %17 = "comp.schedule_func"(%14) ( {
      "gpu.launch_func"(%c1, %c1, %c1, %c1, %c8, %c1, %15) {kernel = @main_kernel_1::@main_kernel} : (index, index, index, index, index, index, memref<1x8x1xf32, 11>) -> ()
      "comp.schedule_end"() : () -> ()
    }) : (!comp.execenv<ocl:0,(11)>) -> !comp.event<ocl>
    %18 = comp.schedule_read %arg3 from %15 on %14 wait for %17 : (memref<1x8x1xf32>, memref<1x8x1xf32, 11>, !comp.execenv<ocl:0,(11)>, !comp.event<ocl>) -> !comp.event<ocl>
    comp.wait %18 : !comp.event<ocl>
    comp.dealloc %14 %15 : (!comp.execenv<ocl:0,(11)>, memref<1x8x1xf32, 11>) -> ()
    comp.destroy_execenv %14 : !comp.execenv<ocl:0,(11)>
    %19 = comp.create_execenv %arg0 : (!comp.device) -> !comp.execenv<ocl:0,(11)>
    %20 = comp.alloc %19 : (!comp.execenv<ocl:0,(11)>) -> memref<1x10x3xf32, 11>
    %21 = comp.schedule_write %0 to %20 on %19 : (memref<1x10x3xf32>, memref<1x10x3xf32, 11>, !comp.execenv<ocl:0,(11)>) -> !comp.event<ocl>
    comp.wait %21 : !comp.event<ocl>
    %22 = comp.alloc %19 : (!comp.execenv<ocl:0,(11)>) -> memref<2x3x1xf32, 11>
    %23 = comp.schedule_write %arg1 to %22 on %19 : (memref<2x3x1xf32>, memref<2x3x1xf32, 11>, !comp.execenv<ocl:0,(11)>) -> !comp.event<ocl>
    comp.wait %23 : !comp.event<ocl>
    %24 = comp.alloc %19 : (!comp.execenv<ocl:0,(11)>) -> memref<1x8x1xf32, 11>
    %25 = comp.schedule_write %arg3 to %24 on %19 : (memref<1x8x1xf32>, memref<1x8x1xf32, 11>, !comp.execenv<ocl:0,(11)>) -> !comp.event<ocl>
    comp.wait %25 : !comp.event<ocl>
    %26 = "comp.schedule_func"(%19) ( {
      "gpu.launch_func"(%c1, %c1, %c1, %c8, %c1, %c1, %20, %22, %24) {kernel = @main_kernel_2::@main_kernel} : (index, index, index, index, index, index, memref<1x10x3xf32, 11>, memref<2x3x1xf32, 11>, memref<1x8x1xf32, 11>) -> ()
      "comp.schedule_end"() : () -> ()
    }) : (!comp.execenv<ocl:0,(11)>) -> !comp.event<ocl>
    %27 = comp.schedule_read %0 from %20 on %19 wait for %26 : (memref<1x10x3xf32>, memref<1x10x3xf32, 11>, !comp.execenv<ocl:0,(11)>, !comp.event<ocl>) -> !comp.event<ocl>
    %28 = comp.schedule_read %arg1 from %22 on %19 wait for %26 : (memref<2x3x1xf32>, memref<2x3x1xf32, 11>, !comp.execenv<ocl:0,(11)>, !comp.event<ocl>) -> !comp.event<ocl>
    %29 = comp.schedule_read %arg3 from %24 on %19 wait for %26 : (memref<1x8x1xf32>, memref<1x8x1xf32, 11>, !comp.execenv<ocl:0,(11)>, !comp.event<ocl>) -> !comp.event<ocl>
    comp.wait %27, %28, %29 : !comp.event<ocl>, !comp.event<ocl>, !comp.event<ocl>
    comp.dealloc %19 %20 : (!comp.execenv<ocl:0,(11)>, memref<1x10x3xf32, 11>) -> ()
    comp.dealloc %19 %22 : (!comp.execenv<ocl:0,(11)>, memref<2x3x1xf32, 11>) -> ()
    comp.dealloc %19 %24 : (!comp.execenv<ocl:0,(11)>, memref<1x8x1xf32, 11>) -> ()
    comp.destroy_execenv %19 : !comp.execenv<ocl:0,(11)>
    return
  }
```
After
```
  func @main(%arg0: !comp.device, %arg1: memref<2x3x1xf32>, %arg2: memref<1x8x3xf32>, %arg3: memref<1x8x1xf32>) {
    %cst = constant 0.000000e+00 : f32
    %c2 = constant 2 : index
    %c4 = constant 4 : index
    %c5 = constant 5 : index
    %c6 = constant 6 : index
    %c7 = constant 7 : index
    %c0 = constant 0 : index
    %c10 = constant 10 : index
    %c3 = constant 3 : index
    %c8 = constant 8 : index
    %c1 = constant 1 : index
    %0 = alloc() : memref<1x10x3xf32>
    %1 = comp.create_execenv %arg0 : (!comp.device) -> !comp.execenv<ocl:0,(11)>
    %2 = comp.alloc %1 : (!comp.execenv<ocl:0,(11)>) -> memref<1x10x3xf32, 11>
    %3 = comp.alloc %1 : (!comp.execenv<ocl:0,(11)>) -> memref<1x8x3xf32, 11>
    %4 = comp.schedule_write %arg2 to %3 on %1 : (memref<1x8x3xf32>, memref<1x8x3xf32, 11>, !comp.execenv<ocl:0,(11)>) -> !comp.event<ocl>
    comp.wait %4 : !comp.event<ocl>
    %5 = comp.alloc %1 : (!comp.execenv<ocl:0,(11)>) -> memref<1x8x1xf32, 11>
    %6 = comp.schedule_write %arg3 to %5 on %1 : (memref<1x8x1xf32>, memref<1x8x1xf32, 11>, !comp.execenv<ocl:0,(11)>) -> !comp.event<ocl>
    comp.wait %6 : !comp.event<ocl>
    %7 = comp.alloc %1 : (!comp.execenv<ocl:0,(11)>) -> memref<2x3x1xf32, 11>
    %8 = comp.schedule_write %arg1 to %7 on %1 : (memref<2x3x1xf32>, memref<2x3x1xf32, 11>, !comp.execenv<ocl:0,(11)>) -> !comp.event<ocl>
    comp.wait %8 : !comp.event<ocl>
    %9 = "comp.schedule_func"(%1) ( {
      "gpu.launch_func"(%c1, %c1, %c1, %c1, %c10, %c3, %2) {kernel = @main_kernel::@main_kernel} : (index, index, index, index, index, index, memref<1x10x3xf32, 11>) -> ()
      "comp.schedule_end"() : () -> ()
    }) : (!comp.execenv<ocl:0,(11)>) -> !comp.event<ocl>
    %10 = "comp.schedule_func"(%1) ( {
      "gpu.launch_func"(%c1, %c1, %c1, %c1, %c8, %c3, %3, %2) {kernel = @main_kernel_0::@main_kernel} : (index, index, index, index, index, index, memref<1x8x3xf32, 11>, memref<1x10x3xf32, 11>) -> ()
      "comp.schedule_end"() : () -> ()
    }) : (!comp.execenv<ocl:0,(11)>) -> !comp.event<ocl>
    %11 = "comp.schedule_func"(%1) ( {
      "gpu.launch_func"(%c1, %c1, %c1, %c1, %c8, %c1, %5) {kernel = @main_kernel_1::@main_kernel} : (index, index, index, index, index, index, memref<1x8x1xf32, 11>) -> ()
      "comp.schedule_end"() : () -> ()
    }) : (!comp.execenv<ocl:0,(11)>) -> !comp.event<ocl>
    %12 = "comp.schedule_func"(%1) ( {
      "gpu.launch_func"(%c1, %c1, %c1, %c8, %c1, %c1, %2, %7, %5) {kernel = @main_kernel_2::@main_kernel} : (index, index, index, index, index, index, memref<1x10x3xf32, 11>, memref<2x3x1xf32, 11>, memref<1x8x1xf32, 11>) -> ()
      "comp.schedule_end"() : () -> ()
    }) : (!comp.execenv<ocl:0,(11)>) -> !comp.event<ocl>
    %13 = comp.schedule_read %arg1 from %7 on %1 wait for %12 : (memref<2x3x1xf32>, memref<2x3x1xf32, 11>, !comp.execenv<ocl:0,(11)>, !comp.event<ocl>) -> !comp.event<ocl>
    %14 = comp.schedule_read %arg3 from %5 on %1 wait for %12 : (memref<1x8x1xf32>, memref<1x8x1xf32, 11>, !comp.execenv<ocl:0,(11)>, !comp.event<ocl>) -> !comp.event<ocl>
    %15 = comp.schedule_read %arg2 from %3 on %1 wait for %12 : (memref<1x8x3xf32>, memref<1x8x3xf32, 11>, !comp.execenv<ocl:0,(11)>, !comp.event<ocl>) -> !comp.event<ocl>
    %16 = comp.schedule_read %0 from %2 on %1 wait for %12 : (memref<1x10x3xf32>, memref<1x10x3xf32, 11>, !comp.execenv<ocl:0,(11)>, !comp.event<ocl>) -> !comp.event<ocl>
    comp.wait %13, %14, %15, %16 : !comp.event<ocl>, !comp.event<ocl>, !comp.event<ocl>, !comp.event<ocl>
    comp.dealloc %1 %7 : (!comp.execenv<ocl:0,(11)>, memref<2x3x1xf32, 11>) -> ()
    comp.dealloc %1 %5 : (!comp.execenv<ocl:0,(11)>, memref<1x8x1xf32, 11>) -> ()
    comp.dealloc %1 %3 : (!comp.execenv<ocl:0,(11)>, memref<1x8x3xf32, 11>) -> ()
    comp.dealloc %1 %2 : (!comp.execenv<ocl:0,(11)>, memref<1x10x3xf32, 11>) -> ()
    comp.destroy_execenv %1 : !comp.execenv<ocl:0,(11)>
    return
  }
```
3. OpenCL and Vulkan backends now share the same comp dialect representation, only memref memorySpace is different.
    https://github.com/plaidml/plaidml/issues/1413

4. adds multi command buffer support to vulkan backend, 
    Vulkan resnet50 kernel execute time changed from 75ms to 83ms due to overheads of multi-commandBuffer.
    Multi command buffer support fixed 12 keras backend tests which were modifing buffer on host side before Vulkan commandBuffer returns, https://github.com/plaidml/plaidml/issues/1414 is not needed.
    Memory transfer using memory map not vkCmd*

jbruestle commented:
Looks good overall.  Two high level points:  1) It would be nice to have a separate PR for the GPU->COMP lowering optimizations and the Vulkan changes, however I tried to pull out just the Vulkan changes and that broke, not sure why exactly (see https://github.com/plaidml/plaidml/pull/1585).  I'm OK landing these as one PR if we need to, but I'd like to understand why better.  2) The changes in GPU->COMP lowering are basically mixing optimization and conversion/lowering into one block of code.  In general the design is more composable if we make conversions as simple as possible and then make optimizations happen afterwards (or before).  Again, I don't think that should necessarily block this PR, but just wanted to comment for future reference.

YangleiZouIntel commented:
> Looks good overall. Two high level points: 1) It would be nice to have a separate PR for the GPU->COMP lowering optimizations and the Vulkan changes, however I tried to pull out just the Vulkan changes and that broke, not sure why exactly (see #1585). I'm OK landing these as one PR if we need to, but I'd like to understand why better. 2) The changes in GPU->COMP lowering are basically mixing optimization and conversion/lowering into one block of code. In general the design is more composable if we make conversions as simple as possible and then make optimizations happen afterwards (or before). Again, I don't think that should necessarily block this PR, but just wanted to comment for future reference.

Thanks Jeremy for the note. 

For (2) I have split the code in this PR in two passes (one before gpu_to_comp and one after gpu_to_comp) and some changes in gpu_to_comp pass, please see https://github.com/plaidml/plaidml/pull/1590 . I will add a separate PR for vulkan. 

For (1) the current Vulkan and OpenCL backends have different comp representation and I changed gpu_to_comp pass in this PR so the vulkan code in #1585 itself will broke if the changes in gpu_to_comp pass in this PR is not applied. With the changes in gpu_to_comp pass, the current Vulkan and OpenCL backends will share the same comp dialect representation, only the buffer memory space and execEnv will be different.

YangleiZouIntel commented:
Closed since this PR is splited into https://github.com/plaidml/plaidml/pull/1590  and https://github.com/plaidml/plaidml/pull/1605
